### PR TITLE
Removes HasTwo by making Filesystem (privately) clonable

### PR DESCRIPTION
**Pull request by nobbele**
Like said in #1053, this removes the ugly tuple borrow

### CODE REVIEW

File: src/filesystem.rs [hunk 1]
```diff
@@ -34,11 +34,12 @@
 
 use crate::{
     conf,
-    vfs::{self, VFS},
+    vfs::{self, OverlayFS, VFS},
     Context, GameError, GameResult,
 };
 use directories::ProjectDirs;
-use std::{env, io, io::SeekFrom, path};
+use parking_lot::Mutex;
```
Comment (PSteinhaus) [inline src/filesystem.rs:41]: why use `parking_lot::Mutex` instead of a standard mutex?

File: src/filesystem.rs [hunk 2]
```diff
@@ -47,13 +48,29 @@ const CONFIG_NAME: &str = "/conf.toml";
...
+        Filesystem {
+            vfs: self.vfs.clone(),
+            resources_dir: self.resources_dir.clone(),
+            zip_dir: self.zip_dir.clone(),
+            user_config_dir: self.user_config_dir.clone(),
+            user_data_dir: self.user_data_dir.clone(),
+        }
+    }
+}
+
```
Comment (PSteinhaus) [inline src/filesystem.rs:73]: Why call this `PrivClone` instead of just implementing `Clone` for Filesystem?

**Review by PSteinhaus — COMMENTED** [review]
It's a bit of a workaround, but I agree that it's the better workaround, compared to `HasTwo`.

File: src/filesystem.rs (hunk 1 above)
Comment (nobbele) [reply src/filesystem.rs:41]: std mutex gained good performance on linux only recently but parking_lot still has better "low contention" performance, I don't expect mutexes to ever block (realized an issue here, mutexes won't work in wasm) so there will always be zero contention. And I always see parking_lot being recommended over std in general but maybe it doesn't make sense to bring in another dependency (and even if we did, it should probably be applied to all mutexes in that case)

Comment (nobbele) [reply src/filesystem.rs:41]: actually nvm parking_lot mutexes do work in wasm as long as they don't block, which they won't if there are no threads, which there aren't in wasm (says some github comment, didn't verify)

Comment (nobbele) [reply src/filesystem.rs:41]: I'll switch to std mutex for now though, we can decide on that later

Comment (PSteinhaus) [reply src/filesystem.rs:41]: > actually nvm parking_lot mutexes do work in wasm as long as they don't block, which they won't if there are no threads, which there aren't in wasm (says some github comment, didn't verify). Do std mutexes work there though?

Looking it up quickly it seems they do. I'd say we stick with `std` until we get a good reason to switch and then switch everywhere at once, yes.

File: src/filesystem.rs (hunk 2 above)
Comment (PSteinhaus) [reply src/filesystem.rs:73]: Hah. That makes sense ^^
Hmm, maybe we should add some kind of reminder to the next 0.9 milestone for this? If we didn't change it by then then I'd just replace this with `Clone`.

For now we can keep it though. But please do add that comment, that's a good idea.

**Comment by nobbele** [comment]
uh did I break it?

**Pull request closed without merging**